app/src/main/java/io/ooni/mk/nuvolari/android/MainActivity.java

```diff
@@ -0,0 +1,68 @@
+package io.ooni.mk.nuvolari.android;
+
+import android.os.Handler;
+import android.os.Looper;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.v7.app.AppCompatActivity;
+import android.os.Bundle;
+import android.util.Log;
+import android.widget.Toast;
+
+import io.ooni.mk.nuvolari.libndt7.Client;
+import io.ooni.mk.nuvolari.libndt7.Measurement;
+import io.ooni.mk.nuvolari.libndt7.Settings;
+
+public class MainActivity extends AppCompatActivity {
+
+    @NonNull
+    private static final String TAG = "MainActivity";
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        Settings settings = new Settings();
+        settings.hostname = "35.235.104.27";
+        settings.port = 443;
+        settings.disableTLS = false;
+        settings.skipTLSVerify = true;
+        settings.download.duration = 1;
+        settings.download.adaptive = true;
+        Client client = new MyClient(settings);
+
+        if (!client.runDownload()) {
+            Toast.makeText(this, "runDownload failed", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    private class MyClient extends Client {
+        private MyClient(@NonNull Settings settings) {
+            super(settings);
+        }
+
+        @Override
+        public void onLogInfo(@Nullable String message) {
+            System.out.println(message);
+        }
+
+        @Override
+        public void onError(@Nullable String error) {
+            System.out.println("Error: " + error);
+
+            new Handler(Looper.getMainLooper()).post(() ->
```

Comment:
Uh, so this is the idiomatic way of sending a message to the UI thread, isn't it?

Reply:
(I see why you asked for Java 8, btw)

Reply:
There are a few different ways these days but this is the built-in android way